src/metric_executor.cpp

```diff
@@ -0,0 +1,75 @@
+// metricq-source-cmd
+// Copyright (C) 2023 ZIH, Technische Universitaet Dresden, Federal Republic of Germany
+//
+// All rights reserved.
+//
+// This file is part of metricq-source-cmd.
+//
+// metricq-source-cmd is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// metricq-source-cmd is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with metricq-source-cmd.  If not, see <http://www.gnu.org/licenses/>.
+#include "metric_executor.hpp"
+
+#include <metricq/logger/nitro.hpp>
+
+#include <metricq/types.hpp>
+
+#include <cstdio>
+
+using Log = metricq::logger::nitro::Log;
+
+MetricExecutor::MetricExecutor(metricq::Metric<metricq::Source>& metric, metricq::Duration interval,
+                               const std::string& command, const std::string& unit,
+                               asio::io_service& io_service)
+: interval(interval), timer_(io_service), command_(command), unit_(unit), metric_(metric)
+{
+    metric_.metadata.unit(unit_);
+    metric_.metadata.rate(1.0 / (interval.count() * 1e-9));
+    metric_.metadata["description"] = command_;
+}
+
+MetricExecutor::~MetricExecutor()
+{
+    timer_.cancel();
+}
+
+metricq::Timer::TimerResult MetricExecutor::timeout_cb(std::error_code)
+{
+    Log::debug() << "sending metrics...";
+    double value = execute_command();
+    auto current_time = metricq::Clock::now();
+    metric_.send({ current_time, value });
+    return metricq::Timer::TimerResult::repeat;
+}
+
+void MetricExecutor::start()
+{
+    timer_.start([this](auto err) { return this->timeout_cb(err); }, interval);
+}
+
+double MetricExecutor::execute_command()
+{
+    // FROM https://stackoverflow.com/a/478960
+    std::array<char, 128> buffer{};
+    std::string result;
+    std::unique_ptr<FILE, decltype(&pclose)> pipe(popen(command_.c_str(), "r"), pclose);
+    if (!pipe)
+    {
+        throw std::runtime_error("popen() failed!");
+    }
+    while (fgets(buffer.data(), buffer.size(), pipe.get()) != nullptr)
```

Comment:
There's https://think-async.com/Asio/asio-1.24.0/doc/asio/reference/readable_pipe.html in asio. Might be a good idea to use that.

Reply:
Also, this whole function, though copied from StackOverflow, feels dirty